tests/components/matrix/test_commands.py

```diff
@@ -0,0 +1,180 @@
+            and self.room_id not in SUBSET_ROOMS
+        ):
+            return None
+        return {
+            "sender": "@SomeUser:example.com",
+            "room": self.room_id,
+        } | self.expected_event_data_extra
+
+
+room_message_base = partial(
+    RoomMessageText,
+    formatted_body=None,
+    format=None,
+    source={
+        "event_id": "fake_event_id",
+        "sender": "@SomeUser:example.com",
+        "origin_server_ts": 123456789,
+    },
+)
+word_command_global = partial(
+    CommandTestParameters,
+    room_message=room_message_base(body="!WordTrigger arg1 arg2"),
+    expected_event_data_extra={
+        "command": "WordTriggerEventName",
+        "args": ["arg1", "arg2"],
+    },
+)
+expr_command_global = partial(
+    CommandTestParameters,
+    room_message=room_message_base(body="My name is FakeName"),
+    expected_event_data_extra={
+        "command": "ExpressionTriggerEventName",
+        "args": {"name": "FakeName"},
+    },
+)
+word_command_subset = partial(
+    CommandTestParameters,
+    room_message=room_message_base(body="!WordTriggerSubset arg1 arg2"),
+    expected_event_data_extra={
+        "command": "WordTriggerSubsetEventName",
+        "args": ["arg1", "arg2"],
+    },
+)
+expr_command_subset = partial(
+    CommandTestParameters,
+    room_message=room_message_base(body="Your name is FakeName"),
+    expected_event_data_extra={
+        "command": "ExpressionTriggerSubsetEventName",
+        "args": {"name": "FakeName"},
+    },
+)
+# Messages without commands should trigger nothing
+fake_command_global = partial(
+    CommandTestParameters,
+    room_message=room_message_base(body="This is not a real command!"),
+    expected_event_data_extra=None,
+)
+# Valid commands sent by the bot user should trigger nothing
+self_command_global = partial(
+    CommandTestParameters,
+    room_message=room_message_base(
+        body="!WordTrigger arg1 arg2",
+        source={
+            "event_id": "fake_event_id",
+            "sender": TEST_MXID,
+            "origin_server_ts": 123456789,
+        },
+    ),
+    expected_event_data_extra=None,
+)
+
+
+@pytest.mark.parametrize(
+    "command_params",
+    chain(
+        (word_command_global(room_id) for room_id in ALL_ROOMS),
+        (expr_command_global(room_id) for room_id in ALL_ROOMS),
+        (word_command_subset(room_id) for room_id in SUBSET_ROOMS),
+        (expr_command_subset(room_id) for room_id in SUBSET_ROOMS),
+    ),
+)
+async def test_commands(
+    hass: HomeAssistant,
+    matrix_bot: MatrixBot,
+    command_events: list[Event],
+    command_params: CommandTestParameters,
+):
+    """Test that the configured commands are used correctly."""
+    room = MatrixRoom(room_id=command_params.room_id, own_user_id=matrix_bot._mx_id)
+
+    await hass.async_start()
+    assert len(command_events) == 0
+    await matrix_bot._handle_room_message(room, command_params.room_message)
```

Comment:
We prefer patching the library and setting up the integration, then calling the notify service and assert mock call args. We shouldn't interact with integration or client details in the tests.

https://developers.home-assistant.io/docs/development_testing#writing-tests-for-integrations

Reply:
In test_notify, that's exactly what's being done for the purpose of actually testing that the underlying implementation detail is working as expected in the context of an automation using the notify service to send a message.

However, using the notify service like that is actively *incorrect* for the purpose of testing commands: the notify service will cause a message to be sent by the user that the bot itself is logged in as.

The bot is *not* supposed to respond to otherwise-valid commands sent by its own user (whether in this current session as used by home assistant, or by another login on another device).

One of the new test cases even explicitly checks for this.

In summary:
`_handle_send_message` is therefore being invoked directly because:
- we test its proper behavior with the notify service in a different set of unit tests
- using it is a convenience over having to completely re-implement sending a message within the test code just to be able to send a message as someone else (when that implementation would look nearly the same as the implementation detail, for good reason)
- also prevent us from having to unnecessarily mock away the *other* implementation detail of `AsyncClient` any further than it already has been
  - `MockAsyncClient` in `conftest` already mocks away the lower-level IO logic performed by the 3rd party dependency `matrix-nio`, but further mocking the actual result of sending a message would make it unsafe (in my view as the actual maintainer of `matrix-nio`) since it'd obscure any changes that occur in its API

I'm not against trying to improve the test, just the opposite, and I appreciate you taking a look! But I'd need to find a way to preserve the total effectiveness/correctness of the tests while doing so, and the notify service simply would cause it to become incorrect: any ideas anyone may have on this are very welcome!

Reply:
I looked at the integration a bit more. `_handle_room_message` is a callback passed to `client.add_event_callback`. So instead of knowing about an integration detail we should access the callback via the `mock_client.add_event_callback.call_args` attribute.

Please refactor away all direct usage of `matrix_bot` and use `mock_client` as needed instead.

Reply:
I recommend autospeccing the mock instead and setting return values on the mock methods that need that, eg the methods that should return responses. An autospecced mock will have all real methods mocked and using other mock methods will raise.

Currently we return a custom class that inherit the real class as mock.